README.md
# NAME

Mojo::UserAgent::Mockable - A Mojo User-Agent that can record and play back requests without Internet connectivity, similar to LWP::UserAgent::Mockable

# VERSION

version 1.56

# SYNOPSIS

    my $ua = Mojo::UserAgent::Mockable->new( mode => 'record', file => '/path/to/file' );
    my $tx = $ua->get($url);

    # Then later...
    my $ua = Mojo::UserAgent::Mockable->new( mode => 'playback', file => '/path/to/file' );
    
    my $tx = $ua->get($url); 
    # This is the same content as above. The saved response is returned, and no HTTP request is
    # sent to the remote host.
    my $reconstituted_content = $tx->res->body;

# ATTRIBUTES

## mode

Mode to operate in.  One of:

- passthrough

    Operates like [Mojo::UserAgent](https://metacpan.org/pod/Mojo::UserAgent) in all respects. No recording or playback happen.

- record

    Records all transactions made with this instance to the file specified by ["file"](#file).

- playback

    Plays back transactions recorded in the file specified by ["file"](#file)

- lwp-ua-mockable

    Works like [LWP::UserAgent::Mockable](https://metacpan.org/pod/LWP::UserAgent::Mockable). Set the LWP\_UA\_MOCK environment variable to 'playback', 
    'record', or 'passthrough', and the LWP\_UA\_MOCK\_FILE environment variable to the recording file.

## file

File to record to / play back from.

## unrecognized

What to do on an unexpected request.  One of:

- exception

    Throw an exception (i.e. die).

- null

    Return a response with empty content

- fallback

    Process the request as if this instance were in "passthrough" mode and perform the HTTP request normally.

## ignore\_headers

Request header names to ignore when comparing a request made with this class to a stored request in 
playback mode. Specify 'all' to remove any headers from consideration. By default, the 'Connection',
'Host', 'Content-Length', and 'User-Agent' headers are ignored.

## ignore\_body

Ignore the request body entirely when comparing a request made with this class to a stored request 
in playback mode.

## ignore\_userinfo 

Ignore the userinfo portion of the request URL's when comparing a request to a potential counterpart in playback mode.

## request\_normalizer

Optional subref. This is for when the requests require a more nuanced comparison (although it will
be used in conjunction with the previous attributes).

The subref takes two parameters: the current Mojo::Message::Request and the recorded one. The subref
should modify these request objects in-place so that they match each other for the parts where your
code doesn't care, e.g. set an id or timestamp to the same value in both requests.

The return value is ignored, so a typical subref to ignore differences in any numerical id parts of
the query path could look like this

    request_normalizer => sub {
        my ($req, $recorded_req) = @_;
        for ($req, $recorded_req) {
            $_->url->path( $_->url->path =~ s|/\d+\b|/123|gr );
        }
    },

# METHODS

## save

In record mode, save the transaction cache to the file specified by ["file"](#file) for later playback.

# THEORY OF OPERATION

## Recording mode

For the life of a given instance of this class, all transactions made using that instance will be 
serialized and stored in memory.  When the instance goes out of scope, or at any time  ["save"](#save) is 
called, the transaction cache will be written to the file specfied by ["file"](#file) in JSON format. 
Transactions are stored in the cache in the order they were made.

## Playback mode

When this class is instantiated, the instance will read the transaction cache from the file 
specified by ["file"](#file). When a request is first made using the instance, if the request matches 
that of the first transaction in the cache, the request URL will be rewritten to that of the local 
host, and the response from the first stored transaction will be returned to the caller. Each 
subsequent request will be handled similarly, and requests must be made in the same order as they 
were originally made, i.e. if orignally the request order was A, B, C, with responses A', B', C',
requests in order A, C, B will NOT return responses A', C', B'. Request A will correctly return 
response A', but request C will trigger an error (behavior configurable by the ["unrecognized"](#unrecognized)
option).

### Request matching

Before comparing the current request with the recorded one, the requests are normalized using the
subref in the request\_normalizer attribute. The default is no normalization. See above for how to
use it.

Two requests are considered to be equivalent if they have the same URL (order of query parameters
notwithstanding), the same body content, and the same headers.

You may also exclude headers from consideration by means of the ["ignore\_headers"](#ignore_headers) attribute. Or,
you may excluse the request body from consideration by means of the ["ignore\_body"](#ignore_body) attribute.

# CAVEATS

## Encryption

The playback file generated by this module is unencrypted JSON.  Treat the playback file as if 
its contents were being transmitted over an unsecured channel.

## Local application server

Using this module against a local app, e.g.: 

    my $app = Mojolicious->new;
    ...

    my $ua = Mojo::UserAgent::Mockable->new;
    $ua->server->app($app);

Doesn't work, because in playback mode, requests are served from an internal Mojolicious instance.
So if you blow that away, the thing stops working, natch.  You should instead instantiate 
[Mojo::Server::Daemon](https://metacpan.org/pod/Mojo::Server::Daemon) and connect to the app via the server's URL, like so:

    use Mojo::Server::Daemon;
    use Mojo::IOLoop;

    my $app = Mojolicious->new;
    $app->routes->any( ... );

    my $daemon = Mojo::Server::Daemon->new(
        app => $app, 
        ioloop => Mojo::IOLoop->singleton,
        silent => 1,
    );
    
    my $listen = q{http://127.0.0.1};
    $daemon->listen( [$listen] )->start;
    my $port = Mojo::IOLoop->acceptor( $daemon->acceptors->[0] )->port;
    my $url  = Mojo::URL->new(qq{$listen:$port})->userinfo('joeblow:foobar');
    
    my $output_file = qq{/path/to/file.json};
    
    my $mock = Mojo::UserAgent::Mockable->new(ioloop => Mojo::IOLoop->singleton, mode => 'record', file => $output_file);
    my $tx = $mock->get($url);

## Mojolicious::Lite

You will often see tests written using [Mojolicious::Lite](https://metacpan.org/pod/Mojolicious::Lite) like so:

    use Mojolicious::Lite;

    get '/' => sub { ... };

    post '/foo' => sub { ... };

And then, further down:

    my $ua = Mojo::UserAgent->new;

    is( $ua->get('/')->res->text, ..., 'Text OK' );
Or:

    use Test::Mojo;
    my $t = Test::Mojo->new;
    $t->get_ok('/')->status_is(200)->text_is( ... );

And this is all fine. Where it stops being fine is when you have Mojo::UserAgent::Mockable on board:

    use Mojolicious::Lite;

    get '/' => sub { ... };

    post '/foo' => sub { ... };
    
    use Test::Mojo;
    my $t = Test::Mojo->new;
    my $mock = Mojo::UserAgent::Mockable->new( mode => 'playback', file => ... );
    $t->get_ok('/')->status_is(200)->text_is( ... );

Mojolicious::Lite will replace the current UA's internal application server's application instance 
(["app" in Mojo::UserAgent::Server](https://metacpan.org/pod/Mojo::UserAgent::Server#app)) with the Mojolicious::Lite application.  This will break the 
playback functionality, as this depends on a custom Mojolicious application internal to the module.
Instead, define your application in a separate package (not necessarily a separate file), like so:

    package MyApp;
    use Mojolicious::Lite;
    get '/' => sub { ... };
    post '/foo' => sub { ... };

    # Actual test application
    package main;

    use Mojo::UserAgent::Mockable;
    use Mojo::Server::Daemon;
    use Mojo::IOLoop;
    use Test::Mojo;

    $app->routes->get('/' => sub { ... });
    $app->routes->post('/foo' => sub { ... });

    my $daemon = Mojo::Server::Daemon->new(
        app    => $app,
        ioloop => Mojo::IOLoop->singleton,
        silent => 1,
    );

    my $listen = q{http://127.0.0.1};
    $daemon->listen( [$listen] )->start;
    my $port = Mojo::IOLoop->acceptor( $daemon->acceptors->[0] )->port;
    my $url  = Mojo::URL->new(qq{$listen:$port})->userinfo('joeblow:foobar');

    my $mock = Mojo::UserAgent::Mockable->new(ioloop => Mojo::IOLoop::singleton, mode => playback, file => ... );
    my $t = Test::Mojo->new;
    $t->ua($mock);
    $mock->get_ok($url->clone->path('/'))->status_is(200)->text_is( ... );

You can also do the following (as seen in t/030\_basic\_authentication.t):

    use Mojolicious;
    use Mojo::Server::Daemon;
    use Mojo::IOLoop;

    my $app = Mojolicious->new;
    $app->routes->get('/' => sub { ... });
    $app->routes->post('/foo' => sub { ... });

    my $daemon = Mojo::Server::Daemon->new(
        app    => $app,
        ioloop => Mojo::IOLoop->singleton,
        silent => 1,
    );

    my $listen = q{http://127.0.0.1};
    $daemon->listen( [$listen] )->start;
    my $port = Mojo::IOLoop->acceptor( $daemon->acceptors->[0] )->port;
    my $url  = Mojo::URL->new(qq{$listen:$port})->userinfo('joeblow:foobar');

    my $mock = Mojo::UserAgent::Mockable->new(ioloop => Mojo::IOLoop::singleton, mode => playback, file => ... );
    my $t = Test::Mojo->new;
    $t->ua($mock);
    $t->get_ok('/')->status_is(200)->content_is( ... );

## Events

The following transaction level events will not be emitted during playback:

- pre\_freeze
=item post\_freeze
=item resume

# SEE ALSO

- [Mojo::UserAgent](https://metacpan.org/pod/Mojo::UserAgent) 
The class being mocked (but not derided, because the whole Mojo thing is really quite clever)
=item \* [Mojo::Transaction::HTTP](https://metacpan.org/pod/Mojo::Transaction::HTTP) 
Where the magic happens

# CONTRIBUTORS

Mike Eve [https://github.com/ungrim97](https://github.com/ungrim97)

Phineas J. Whoopee [https://github.com/antoniel123](https://github.com/antoniel123)

Marc Murray [https://github.com/marcmurray](https://github.com/marcmurray)

Steve Wagner `<truroot at gmail.com>`

Joel Berger `<joel.a.berger at gmail.com>`

Dan Book `<grinnz at grinnz.com>`

Stefan Adams  `<stefan@borgia.com>`

Mohammad Anwar `mohammad.anwar@yahoo.com`

Johan Lindstrom `johanl@cpan.org`

Everyone on #mojo on irc.perl.org

# AUTHOR

Kit Peters <popefelix@gmail.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2018 by Kit Peters.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
{{$NEXT}}

- #15: Fixed issue where save wouldn't work if called during global destruction

1.55      2018-11-21 22:05:33-06:00 America/Chicago

- Mojo::UserAgent::Mockable::Request::Compare compares GET parameters without regard to parameter order.
- Removed dependency on Mojolicious::Quick in favor of Mojolicious::Lite

1.53      2017-02-27 15:58:36-06:00 America/Chicago

- Fixed dependencies screwed up in v1.52
- Fixed issue #18 where tests failed due updates to the V&A site.

1.52      2017-02-27 10:21:36-06:00 America/Chicago

- Updated dependencies - no functional changes

1.51      2017-02-24 14:49:29-06:00 America/Chicago

- Sessions going through a CONNECT proxy can now be played back properly (issue #9). 
- Added option to disable userinfo comparision
- Mojo::Util::slurp has been removed in favor of Path::Tiny::slurp in tests

1.42      2016-07-29 10:35:40-05:00 America/Chicago

- Added minimum Mojlicious version of 6. Renamed all numbered tests (save one) to unnumbered tests.

1.41      2016-07-28 10:23:13-05:00 America/Chicago

- Added quota checking for all calls to random.org. 

1.40      2016-07-27 18:34:19-05:00 America/Chicago

- Updated to work with Mojolicious 7.0

1.30      2016-07-05 10:34:05-05:00 America/Chicago

- Added support for normalizing requests. Patch courtesy of Johan Lindstrom 
  <johanl@cpan.org>. 

1.24      2016-04-20 15:52:26-05:00 America/Chicago

- Fix broken dist.ini 

1.23      2016-04-13 14:44:41-05:00 America/Chicago

- Update tests to deal with failures due to missing / old versions of 
  IO::Socket::SSL 

1.22      2016-04-13 13:03:44-05:00 America/Chicago

- Auto-add "provides" to META.* by means of 
  Dist::Zilla::Plugin::MetaProvides::Package. Submitted by Mohammad Anwar
  (https://github.com/Manwar).

1.21      2015-12-28 11:54:21-06:00 America/Chicago

- Version bump with fixed Changes file and two digit version.

1.2       2015-12-23 11:23:24-06:00 America/Chicago

- Added 'ignore_body' option

1.11      2015-12-01 15:33:54-08:00 America/Los_Angeles

- Added missing 'use Mojolicious'. 

1.10      2015-11-23 22:00:19-06:00 America/Chicago

- If we played back a transaction containing an instance of Mojo::URL, that URL 
  would be rewritten during playback. No big deal, except when you want to then
  use that URL for something else.

1.04      2015-11-16 10:51:00-05:00 America/Chicago 
 
- Updated unit tests again to completely remove dependency on PK value from Object 1
  from Victoria and Albert Museum.

1.03      2015-11-10 10:51:00-05:00 America/Chicago 

- Updated unit tests to deal with changed PK value for object 1 from Victoria and 
  Albert Museum.

1.02      2015-10-22 09:33:02-05:00 America/Chicago

- Removed all dependencies on File::Slurp, replaced with File::Slurper or
  Mojo::Util. See 
  http://blogs.perl.org/users/leon_timmermans/2015/08/fileslurp-is-broken-and-wrong.html

1.01      2015-10-12 09:33:02-05:00 America/Chicago

- Tests were referencing a file local to my work machine that was not in the
  distribution. Oops.

1.00      2015-10-09 12:50:20-05:00 America/Chicago

- Initial revision


lib/Mojo/UserAgent/Mockable/Request/Compare.pm
use 5.014;

package Mojo::UserAgent::Mockable::Request::Compare;

# VERSION

# ABSTRACT: Helper class for Mojo::UserAgent::Mockable that compares two Mojo::Message::Request instances

=attr compare_result

The result of the last compare operation.  It is only populated when two requests B<do not> match.

=attr ignore_userinfo 

Set this to a true value to ignore a mismatch in the L<userinfo|Mojo::URL/userinfo> portion of a transaction URL.

=attr ignore_body

Set this to a true value to ignore a mismatch in the bodies of the two compared transactions

=method compare

Compare two instances of L<Mojo::Message::Request>.

=cut

use Carp;
use Mojo::Base -base;
use Mojo::URL;
use Safe::Isa qw{$_isa};

has compare_result => '';
has ignore_headers => sub { [] };
has ignore_body => '';
has ignore_userinfo => '';

sub compare {
    my ($self, $r1, $r2) = @_;
    
    if (!$r1->$_isa('Mojo::Message::Request')) {
        my $reftype = ref $r1;
        croak qq{Cannot compare $reftype to Mojo::Message::Request};
    }
    
    if (!$r2->$_isa('Mojo::Message::Request')) {
        my $reftype = ref $r2;
        croak qq{Cannot compare Mojo::Message::Request to $reftype};
    }

    if ( $r1->method ne $r2->method )  {
        $self->compare_result( sprintf q{Method mismatch: got '%s', expected '%s'}, $r1->method, $r2->method );
        return 0;
    }

    if ( !$self->_compare_url( $r1->url, $r2->url ) ) {
        return 0;
    }

    if ( !$self->ignore_body && $r1->body ne $r2->body ) {
        $self->compare_result('Body mismatch');
        return 0;
    }

    if ($self->ignore_headers ne 'all') {
        my $h1 = $r1->headers->to_hash;
        my $h2 = $r2->headers->to_hash;

        for my $header (@{$self->ignore_headers}) {
            delete $h1->{$header};
            delete $h2->{$header};
        }

        if (scalar keys %{$h1} ne scalar keys %{$h2}) {
            $self->compare_result('Header count mismatch');
                return 0;
        }

        for my $header (keys %{$h1}) {
            if (!defined $h2->{$header}) {
                $self->compare_result(qq{Header "$header" mismatch: header not present in both requests.'});
                return 0;
            }

            if ($h1->{$header} ne $h2->{$header}) {
                no warnings qw/uninitialized/;
                $self->compare_result(qq{Header "$header" mismatch: got '$h1->{$header}', expected '$h2->{$header}'});
                return 0;
            }
        }
    }

    $self->compare_result('');
    return 1;
}

sub _compare_url {
    my ($self, $u1, $u2) = @_;

    if (!ref $u1) {
        $u1 = Mojo::URL->new($u1);
    }
    $u1 = $u1->to_abs;

    if (!ref $u2) {
        $u2 = Mojo::URL->new($u2);
    }
    $u2 = $u2->to_abs;

    no warnings qw/uninitialized/;
    for my $key (qw/scheme userinfo host port fragment/) {
        my $ignore = sprintf 'ignore_%s', $key;
        next if $self->can($ignore) && $self->$ignore;

        my $val1 = $u1->$key;
        my $val2 = $u2->$key;
        if ($val1 ne $val2) {
            $self->compare_result(qq{URL $key mismatch: got "$val1", expected "$val2"});
            return 0;
        }
    }
    
    my $p1 = Mojo::Path->new($u1->path);
    my $p2 = Mojo::Path->new($u2->path);
    if ($p1->to_string ne $p2->to_string) {
        my $val1 = $p1->to_string;
        my $val2 = $p2->to_string;
        $self->compare_result(qq{URL path mismatch: got "$val1", expected "$val2"});
        return 0;
    }

    my $q1 = $u1->query->to_hash;
    my $q2 = $u2->query->to_hash;

    if (scalar keys %{$q1} != scalar keys %{$q2}) {
        my $count1 = scalar keys %{$q1};
        my $count2 = scalar keys %{$q2};

        $self->compare_result(qq{URL query mismatch: parameter count mismatch: $count1 != $count2});
        return 0;
    }
    for my $key (keys %{$q1}) {
        my $val1 = $q1->{$key};
        my $val2 = $q2->{$key};

        if ( ref $val2 eq 'ARRAY' ){
            $val1 = join(",", sort { $a cmp $b } @{$val1});
            $val2 = join(",", sort { $a cmp $b } @{$val2});
        }

        if ($val1 ne $val2) {
            $self->compare_result(qq{URL query mismatch: for key "$key", got "$val1", expected "$val2"});
            return 0;
        }
    }
    use warnings qw/uninitialized/;

    return 1;
}

1;
__END__


lib/Mojo/UserAgent/Mockable.pm
use 5.014;

package Mojo::UserAgent::Mockable;

use warnings::register;

use Carp;
use JSON::MaybeXS;
use Mojolicious 7.22;
use Mojo::Base 'Mojo::UserAgent';
use Mojo::Util qw/secure_compare/;
use Mojo::UserAgent::Mockable::Proxy;
use Mojo::UserAgent::Mockable::Serializer;
use Mojo::UserAgent::Mockable::Request::Compare;
use Mojo::JSON;
use Scalar::Util;

# ABSTRACT: A Mojo User-Agent that can record and play back requests without Internet connectivity, similar to LWP::UserAgent::Mockable

=head1 SYNOPSIS

    my $ua = Mojo::UserAgent::Mockable->new( mode => 'record', file => '/path/to/file' );
    my $tx = $ua->get($url);

    # Then later...
    my $ua = Mojo::UserAgent::Mockable->new( mode => 'playback', file => '/path/to/file' );
    
    my $tx = $ua->get($url); 
    # This is the same content as above. The saved response is returned, and no HTTP request is
    # sent to the remote host.
    my $reconstituted_content = $tx->res->body;

=attr mode

Mode to operate in.  One of:

=for :list
= passthrough
Operates like L<Mojo::UserAgent> in all respects. No recording or playback happen.
= record
Records all transactions made with this instance to the file specified by L</file>.
= playback
Plays back transactions recorded in the file specified by L</file>
= lwp-ua-mockable
Works like L<LWP::UserAgent::Mockable>. Set the LWP_UA_MOCK environment variable to 'playback', 
'record', or 'passthrough', and the LWP_UA_MOCK_FILE environment variable to the recording file.

=attr file

File to record to / play back from.

=attr unrecognized

What to do on an unexpected request.  One of:

=for :list
= exception
Throw an exception (i.e. die).
= null
Return a response with empty content
= fallback
Process the request as if this instance were in "passthrough" mode and perform the HTTP request normally.

=attr ignore_headers

Request header names to ignore when comparing a request made with this class to a stored request in 
playback mode. Specify 'all' to remove any headers from consideration. By default, the 'Connection',
'Host', 'Content-Length', and 'User-Agent' headers are ignored.

=attr ignore_body

Ignore the request body entirely when comparing a request made with this class to a stored request 
in playback mode.

=attr ignore_userinfo 

Ignore the userinfo portion of the request URL's when comparing a request to a potential counterpart in playback mode.

=attr request_normalizer

Optional subref. This is for when the requests require a more nuanced comparison (although it will
be used in conjunction with the previous attributes).

The subref takes two parameters: the current Mojo::Message::Request and the recorded one. The subref
should modify these request objects in-place so that they match each other for the parts where your
code doesn't care, e.g. set an id or timestamp to the same value in both requests.

The return value is ignored, so a typical subref to ignore differences in any numerical id parts of
the query path could look like this

    request_normalizer => sub {
        my ($req, $recorded_req) = @_;
        for ($req, $recorded_req) {
            $_->url->path( $_->url->path =~ s|/\d+\b|/123|gr );
        }
    },


=head1 THEORY OF OPERATION

=head2 Recording mode

For the life of a given instance of this class, all transactions made using that instance will be 
serialized and stored in memory.  When the instance goes out of scope, or at any time  L</save> is 
called, the transaction cache will be written to the file specfied by L</file> in JSON format. 
Transactions are stored in the cache in the order they were made.

=head2 Playback mode

When this class is instantiated, the instance will read the transaction cache from the file 
specified by L</file>. When a request is first made using the instance, if the request matches 
that of the first transaction in the cache, the request URL will be rewritten to that of the local 
host, and the response from the first stored transaction will be returned to the caller. Each 
subsequent request will be handled similarly, and requests must be made in the same order as they 
were originally made, i.e. if orignally the request order was A, B, C, with responses A', B', C',
requests in order A, C, B will NOT return responses A', C', B'. Request A will correctly return 
response A', but request C will trigger an error (behavior configurable by the L</unrecognized>
option).

=head3 Request matching

Before comparing the current request with the recorded one, the requests are normalized using the
subref in the request_normalizer attribute. The default is no normalization. See above for how to
use it.

Two requests are considered to be equivalent if they have the same URL (order of query parameters
notwithstanding), the same body content, and the same headers.

You may also exclude headers from consideration by means of the L</ignore_headers> attribute. Or,
you may excluse the request body from consideration by means of the L</ignore_body> attribute.



=head1 CAVEATS

=head2 Encryption

The playback file generated by this module is unencrypted JSON.  Treat the playback file as if 
its contents were being transmitted over an unsecured channel.

=head2 Local application server

Using this module against a local app, e.g.: 

    my $app = Mojolicious->new;
    ...

    my $ua = Mojo::UserAgent::Mockable->new;
    $ua->server->app($app);

Doesn't work, because in playback mode, requests are served from an internal Mojolicious instance.
So if you blow that away, the thing stops working, natch.  You should instead instantiate 
L<Mojo::Server::Daemon> and connect to the app via the server's URL, like so:

    use Mojo::Server::Daemon;
    use Mojo::IOLoop;

    my $app = Mojolicious->new;
    $app->routes->any( ... );

    my $daemon = Mojo::Server::Daemon->new(
        app => $app, 
        ioloop => Mojo::IOLoop->singleton,
        silent => 1,
    );
    
    my $listen = q{http://127.0.0.1};
    $daemon->listen( [$listen] )->start;
    my $port = Mojo::IOLoop->acceptor( $daemon->acceptors->[0] )->port;
    my $url  = Mojo::URL->new(qq{$listen:$port})->userinfo('joeblow:foobar');
    
    my $output_file = qq{/path/to/file.json};
    
    my $mock = Mojo::UserAgent::Mockable->new(ioloop => Mojo::IOLoop->singleton, mode => 'record', file => $output_file);
    my $tx = $mock->get($url);

=head2 Mojolicious::Lite

You will often see tests written using L<Mojolicious::Lite> like so:

    use Mojolicious::Lite;

    get '/' => sub { ... };

    post '/foo' => sub { ... };

And then, further down:

    my $ua = Mojo::UserAgent->new;

    is( $ua->get('/')->res->text, ..., 'Text OK' );
Or:

    use Test::Mojo;
    my $t = Test::Mojo->new;
    $t->get_ok('/')->status_is(200)->text_is( ... );

And this is all fine. Where it stops being fine is when you have Mojo::UserAgent::Mockable on board:

    use Mojolicious::Lite;

    get '/' => sub { ... };

    post '/foo' => sub { ... };
    
    use Test::Mojo;
    my $t = Test::Mojo->new;
    my $mock = Mojo::UserAgent::Mockable->new( mode => 'playback', file => ... );
    $t->get_ok('/')->status_is(200)->text_is( ... );

Mojolicious::Lite will replace the current UA's internal application server's application instance 
(L<Mojo::UserAgent::Server/app>) with the Mojolicious::Lite application.  This will break the 
playback functionality, as this depends on a custom Mojolicious application internal to the module.
Instead, define your application in a separate package (not necessarily a separate file), like so:

    package MyApp;
    use Mojolicious::Lite;
    get '/' => sub { ... };
    post '/foo' => sub { ... };

    # Actual test application
    package main;

    use Mojo::UserAgent::Mockable;
    use Mojo::Server::Daemon;
    use Mojo::IOLoop;
    use Test::Mojo;

    $app->routes->get('/' => sub { ... });
    $app->routes->post('/foo' => sub { ... });

    my $daemon = Mojo::Server::Daemon->new(
        app    => $app,
        ioloop => Mojo::IOLoop->singleton,
        silent => 1,
    );

    my $listen = q{http://127.0.0.1};
    $daemon->listen( [$listen] )->start;
    my $port = Mojo::IOLoop->acceptor( $daemon->acceptors->[0] )->port;
    my $url  = Mojo::URL->new(qq{$listen:$port})->userinfo('joeblow:foobar');

    my $mock = Mojo::UserAgent::Mockable->new(ioloop => Mojo::IOLoop::singleton, mode => playback, file => ... );
    my $t = Test::Mojo->new;
    $t->ua($mock);
    $mock->get_ok($url->clone->path('/'))->status_is(200)->text_is( ... );

You can also do the following (as seen in t/030_basic_authentication.t):

    use Mojolicious;
    use Mojo::Server::Daemon;
    use Mojo::IOLoop;

    my $app = Mojolicious->new;
    $app->routes->get('/' => sub { ... });
    $app->routes->post('/foo' => sub { ... });

    my $daemon = Mojo::Server::Daemon->new(
        app    => $app,
        ioloop => Mojo::IOLoop->singleton,
        silent => 1,
    );

    my $listen = q{http://127.0.0.1};
    $daemon->listen( [$listen] )->start;
    my $port = Mojo::IOLoop->acceptor( $daemon->acceptors->[0] )->port;
    my $url  = Mojo::URL->new(qq{$listen:$port})->userinfo('joeblow:foobar');

    my $mock = Mojo::UserAgent::Mockable->new(ioloop => Mojo::IOLoop::singleton, mode => playback, file => ... );
    my $t = Test::Mojo->new;
    $t->ua($mock);
    $t->get_ok('/')->status_is(200)->content_is( ... );

=head2 Events

The following transaction level events will not be emitted during playback:

=for :list
=item pre_freeze
=item post_freeze
=item resume

=method save

In record mode, save the transaction cache to the file specified by L</file> for later playback.

=head1 SEE ALSO

=for :list
=item * L<Mojo::UserAgent> 
The class being mocked (but not derided, because the whole Mojo thing is really quite clever)
=item * L<Mojo::Transaction::HTTP> 
Where the magic happens

=head1 CONTRIBUTORS

Mike Eve L<https://github.com/ungrim97>

Phineas J. Whoopee L<https://github.com/antoniel123>

Marc Murray L<https://github.com/marcmurray>

Steve Wagner C<< <truroot at gmail.com> >>

Joel Berger C<< <joel.a.berger at gmail.com> >>

Dan Book C<< <grinnz at grinnz.com> >>

Stefan Adams  C<< <stefan@borgia.com> >>

Mohammad Anwar C<< mohammad.anwar@yahoo.com >>

Johan Lindstrom C<< johanl@cpan.org >>

Everyone on #mojo on irc.perl.org

=cut

has 'mode' => 'passthrough';
has 'file';
has 'unrecognized' => 'exception';
has 'request_normalizer';
has '_serializer' => sub { Mojo::UserAgent::Mockable::Serializer->new };
has 'comparator';
has 'ignore_headers' => sub { [] };
has 'ignore_body';
has 'ignore_userinfo';
has '_mode';
has '_current_txn';
has '_compare_result';
has '_non_blocking';

# Internal Mojolicious app that handles transaction playback
has '_app' => sub {
    my $self = shift;
    my $app  = Mojolicious->new;
    $app->routes->any(
        '/*any' => { any => '' } => sub {
            my $c  = shift;
            my $tx = $c->tx;

            my $txn = $self->_current_txn;
            if ($txn) {
                $self->cookie_jar->collect($txn);
                $tx->res( $txn->res );
                $tx->res->headers->header( 'X-MUA-Mockable-Regenerated' => 1 );
                $c->rendered( $txn->res->code );
            }
            else {
                for my $header ( keys %{ $tx->req->headers->to_hash } ) {
                    if ( $header =~ /^X-MUA-Mockable/ ) {
                        my $val = $tx->req->headers->header($header);
                        $tx->res->headers->header( $header, $val );
                    }
                }
                $c->render( text => '' );
            }
        },
    );
    $app;
};

sub new {
    my $class           = shift;
    my $self            = $class->SUPER::new(@_);
    my %comparator_args = (
        ignore_headers  => 'all',
        ignore_body     => $self->ignore_body,
        ignore_userinfo => $self->ignore_userinfo,
    );
    $self->comparator( Mojo::UserAgent::Mockable::Request::Compare->new(%comparator_args) );

    $self->{'_launchpid'} = $$;
    if ( $self->mode eq 'lwp-ua-mockable' ) {
        $self->_mode( $ENV{'LWP_UA_MOCK'} );
        if ( $self->file ) {
            croak qq{Do not specify 'file' when 'mode' is set to 'lwp-ua-mockable'. Use the LWP_UA_MOCK_FILE }
                . q{environment var instead.};
        }
        $self->file( $ENV{'LWP_UA_MOCK_FILE'} );
    }
    elsif ( $self->mode ne 'record' && $self->mode ne 'playback' && $self->mode ne 'passthrough' ) {
        croak q{Invalid mode. Must be one of 'lwp-ua-mockable', 'record', 'playback', or 'passthrough'};
    }
    else {
        $self->_mode( $self->mode );
    }

    if ( $self->mode eq 'playback' ) {
        $self->proxy(undef);
    }

    if ( $self->_mode ne 'passthrough' && !$self->file ) {
        croak qq{Error: You must specify a recording file};
    }

    if ( $self->_mode ne 'passthrough' ) {
        my $mode      = lc $self->_mode;
        my $mode_init = qq{_init_$mode};
        if ( !$self->can($mode_init) ) {
            croak qq{Error: unsupported mode "$mode"};
        }
        return $self->$mode_init;
    }

    return $self;
}

sub proxy {
    my $self = shift;
    return $self->SUPER::proxy unless @_;

    if ( $self->mode eq 'playback' ) {
        return $self->SUPER::proxy( Mojo::UserAgent::Mockable::Proxy->new );
    }
    else {
        return $self->SUPER::proxy(@_);
    }
}

sub save {
    my ( $self, $file ) = @_;
    if ( $self->_mode eq 'record' ) {
        $file ||= $self->file;

        my $transactions = $self->{'_transactions'};
        $self->_serializer->store( $file, @{$transactions} );
    }
    else {
        carp 'save() only works in record mode' if warnings::enabled;
    }
}

sub start {
    my ( $self, $tx, $cb ) = @_;
    if ($cb) {
        $self->_non_blocking(1);
    }
    return $self->SUPER::start( $tx, $cb );
}

sub _init_playback {
    my $self = shift;

    if ( !-e $self->file ) {
        my $file = $self->file;
        croak qq{Playback file $file not found};
    }
    $self->{'_transactions'} = [ $self->_serializer->retrieve( $self->file ) ];
    my $recorded_tx_count = scalar @{ $self->{_transactions} };

    $self->server->app( $self->_app );

    Scalar::Util::weaken($self);
    $self->on(
        start => sub {
            my ( $ua, $tx ) = @_;

            my $port = $self->_non_blocking ? $self->server->nb_url->port : $self->server->url->port;
            my $recorded_tx = shift @{ $self->{'_transactions'} };

            my ( $this_req, $recorded_req ) = $self->_normalized_req( $tx, $recorded_tx );

            if ( $self->comparator->compare( $this_req, $recorded_req ) ) {
                $self->_current_txn($recorded_tx);

                $tx->req->url( $tx->req->url->clone );
                $tx->req->url->host('')->scheme('')->port($port);
            }
            else {
                unshift @{ $self->{'_transactions'} }, $recorded_tx;

                my $result = $self->comparator->compare_result;
                $self->_current_txn(undef);
                if ( $self->unrecognized eq 'exception' ) {
                    croak qq{Unrecognized request: $result};
                }
                elsif ( $self->unrecognized eq 'null' ) {
                    $tx->req->headers->header( 'X-MUA-Mockable-Request-Recognized'      => 0 );
                    $tx->req->headers->header( 'X-MUA-Mockable-Request-Match-Exception' => $result );
                    $tx->req->url->host('')->scheme('')->port($port);
                }
                elsif ( $self->unrecognized eq 'fallback' ) {
                    $tx->on(
                        finish => sub {
                            my $self = shift;
                            $tx->req->headers->header( 'X-MUA-Mockable-Request-Recognized'      => 0 );
                            $tx->req->headers->header( 'X-MUA-Mockable-Request-Match-Exception' => $result );
                        }
                    );
                }
            }
        }
    );

    return $self;
}

sub _normalized_req {
    my $self = shift;
    my ( $tx, $recorded_tx ) = @_;

    my $request_normalizer = $self->request_normalizer or return ( $tx->req, $recorded_tx->req );
    croak("The request_normalizer attribute is not a coderef") if ( ref($request_normalizer) ne "CODE" );

    my $req          = $tx->req->clone;
    my $recorded_req = $recorded_tx->req->clone;
    $request_normalizer->( $req, $recorded_req );    # To be modified in-place

    return ( $req, $recorded_req );
}

sub _init_record {
    my $self = shift;

    Scalar::Util::weaken($self);
    $self->on(
        start => sub {
            my $tx = $_[1];

            if ( $tx->req->proxy ) {

                # HTTP CONNECT - used for proxy
                return if $tx->req->method eq 'CONNECT';

                # If the TX has a connection assigned, then the request is a copy of the request
                # that initiated the proxy connection
                return if $tx->connection;
            }

            $tx->once(
                finish => sub {
                    my $tx = shift;
                    push @{ $self->{'_transactions'} }, $tx;

                    # 15: During global destruction the $tx object may no longer exist
                    # so save now
                    $self->save($self->file);
                },
            );
            1;
        },
    );

    return $self;
}

sub _load_transactions {
    my ($self) = @_;

    my @transactions = $self->_serializer->retrieve( $self->file );
    return \@transactions;
}

1;
